src/Npgsql/NpgsqlConnector.Auth.cs

```diff
@@ -67,19 +69,87 @@ async Task AuthenticateCleartext(string username, bool async)
 
         async Task AuthenticateSASL(List<string> mechanisms, string username, bool async)
         {
-            // At the time of writing PostgreSQL only supports SCRAM-SHA-256
-            if (!mechanisms.Contains("SCRAM-SHA-256"))
-                throw new NpgsqlException("No supported SASL mechanism found (only SCRAM-SHA-256 is supported for now). " +
+            // At the time of writing PostgreSQL only supports SCRAM-SHA-256 and SCRAM-SHA-256-PLUS
+            var supportsSha256 = mechanisms.Contains("SCRAM-SHA-256");
+            var supportsSha256Plus = mechanisms.Contains("SCRAM-SHA-256-PLUS");
+            if (!supportsSha256 && !supportsSha256Plus)
+                throw new NpgsqlException("No supported SASL mechanism found (only SCRAM-SHA-256 and SCRAM-SHA-256-PLUS are supported for now). " +
                                           "Mechanisms received from server: " + string.Join(", ", mechanisms));
-            var mechanism = "SCRAM-SHA-256";
+
+            var mechanism = string.Empty;
+            var cbindFlag = string.Empty;
+            var cbind = string.Empty;
+            var successfulBind = false;
+
+            if (supportsSha256Plus)
+            {
+                // RFC 5929
+                mechanism = "SCRAM-SHA-256-PLUS";
+                // PostgreSQL only supports tls-server-end-point binding
+                cbindFlag = "p=tls-server-end-point";
+                // SCRAM-SHA-256-PLUS depends on using ssl stream, so it's fine
+                var sslStream = (SslStream)_stream;
+                var cbindFlagBytes = Encoding.UTF8.GetBytes($"{cbindFlag},,");
+
+                using var remoteCertificate = new X509Certificate2(sslStream.RemoteCertificate);
+                // Checking for hashing algorithms
+                HashAlgorithm? hashAlgorithm = null;
+                var algorithmName = remoteCertificate.SignatureAlgorithm.FriendlyName;
+                if (algorithmName.IndexOf("sha1", StringComparison.OrdinalIgnoreCase) >= 0 ||
+                    algorithmName.IndexOf("md5", StringComparison.OrdinalIgnoreCase) >= 0 ||
+                    algorithmName.IndexOf("sha256", StringComparison.OrdinalIgnoreCase) >= 0)
+                {
+                    hashAlgorithm = SHA256.Create();
+                }
+                else if (algorithmName.IndexOf("sha384", StringComparison.OrdinalIgnoreCase) >= 0)
+                {
+                    hashAlgorithm = SHA384.Create();
+                }
+                else if (algorithmName.IndexOf("sha512", StringComparison.OrdinalIgnoreCase) >= 0)
+                {
+                    hashAlgorithm = SHA512.Create();
+                }
+                else
+                {
+                    Log.Warn($"Support for signature algorithm {algorithmName} is not yet implemented, falling back to SCRAM-SHA-256");
+                }
+
+                if (hashAlgorithm != null)
+                {
+                    using (hashAlgorithm)
```

Comment:
nit: can use using declaration instead of block (with _ as discard variable)

Reply:
Not sure what you've meant here. The only reason there is a block is because it's possible to not find suitable hashing algorithm. And in this case, we should try with default `SCRAM-SHA-256`.
Now, I'm thinking it should be refactored into 4 methods (other 3 can be inside the first one):

1. `AuthenticateSASL`, which is a main method
2. `AuthenticateScramPlus`, which tries to prepares required params and authenticate with `SCRAM-SHA-256-PLUS` (and returns `false` if it can't)
3. `AuthenticateScram`, which tries to prepare required params and authenticate with `SCRAM-SHA-256` (and returns `false` if it can't)
4. `Authenticate`, which does all the work